nbqa/replace_source.py

```diff
@@ -87,24 +169,116 @@ def main(python_file: "Path", notebook: "Path", notebook_info: NotebookInfo) ->
     """
     notebook_json = json.loads(notebook.read_text())
 
-    pyfile = python_file.read_text()
+    pycells = _get_pycells(python_file)
 
-    pycells: Iterator[str] = iter(pyfile[len(CODE_SEPARATOR) :].split(CODE_SEPARATOR))
     code_cell_number = 0
     new_cells = []
     for cell in notebook_json["cells"]:
         if cell["cell_type"] == "code":
             code_cell_number += 1
-            if code_cell_number not in notebook_info.code_cells_to_ignore:
-                source = _restore_semicolon(
-                    next(pycells), code_cell_number, notebook_info.trailing_semicolons
-                )
-                cell["source"] = _reinstate_magics(
-                    source,
-                    notebook_info.temporary_lines.get(code_cell_number, []),
-                )
-
+            new_source = _get_new_source(code_cell_number, notebook_info, pycells)
+            if new_source is not None:
+                cell["source"] = new_source
         new_cells.append(cell)
 
     notebook_json.update({"cells": new_cells})
     notebook.write_text(f"{json.dumps(notebook_json, indent=1, ensure_ascii=False)}\n")
+
+
+def print_red(message: str) -> str:
+    """
+    Print message in red.
+
+    Parameters
+    ----------
+    message
+        String to print in red.
+    """
+    return f"{RED}{message}{RESET}"
+
+
+def print_green(message: str) -> str:
+    """
+    Print message in green.
+
+    Parameters
+    ----------
+    message
+        String to print in green.
+    """
+    return f"{GREEN}{message}{RESET}"
+
+
+def _print_diff(code_cell_number: int, cell_diff: Iterator[str]) -> None:
+    """
+    Print diff between cells, colouring as appropriate.
+
+    Parameters
+    ----------
+    code_cell_number
+        Current cell number
+    cell_diff
+        Diff between original and new versions of cell.
+    """
+    peek, cell_diff = _peek(cell_diff)
+    if peek is not None:
+        header = f"Cell {code_cell_number}"
+        sys.stdout.write(f"{BOLD}{header}{RESET}\n")
+        sys.stdout.write(f"{'-'*len(header)}\n")
+        for line in cell_diff:
+            if line.startswith("+++") or line.startswith("---"):
+                sys.stdout.write(line)
+            elif line.startswith("+"):
+                sys.stdout.write(print_green(line))
+            elif line.startswith("-"):
+                sys.stdout.write(print_red(line))
+            else:
+                sys.stdout.write(line)
+        sys.stdout.write("\n")
+
+
+def add_newline_to_last_line(cell: List[str]) -> None:
+    """
+    Add trailing newline to last line in cell, as required by difflib.
+
+    Parameters
+    ----------
+    cell
+        Current cell being processed.
+    """
+    if cell:
+        cell[-1] += "\n"
+
+
+def diff(python_file: "Path", notebook: "Path", notebook_info: NotebookInfo) -> None:
+    """
+    Replace :code:`source` code cells of original notebook.
```

Comment:
Correct the function description `Replace :code:`source` code cells of original notebook.`